### PR TITLE
Allow to turn a particular root to read-only mode

### LINKED ISSUE #13390: Allow to turn a particular root to read-only mode
## 📝 Provide a description of the new feature

Currently, we support read-only mode for the editor. The limitation is that you have to put the entire editor into this mode. This is not flexible enough for integrations using multi-root editor. It is easy to come up with cases, where you would need to put only some of the editables into read-only mode (e.g. you don't have permissions to change some of the content but you can still comment on it, etc.).

### PR DESCRIPTION
**Pull request by kbinieda**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Feature: Allow to turn a particular root to read-only mode. Closes #13390

### CODE REVIEW

**Comment by kbinieda** [comment]
Ok, here's initial PoC to visualise the concept how "multi read-only" may be achieved:

This concept is split into 2 separate parts (may even have 2 different PRs):

1. Add information / flag that can be used to distinguish between particular root readonly characteristics:
Currently, we got `isReadOnly` flag for view `EditableElement` component which is helpful for ui layer. Similarly we can add `isReadOnly` flag for its model counterpart (to be exact I've added it in `RootElement` as adding it in `Element` itself may be debatable). The flow of binding is similar between model and view layers, that is:
`editor.isReadOnly -> model or view document.isReadOnly -> editableElement or rootElement.isReadOnly`.

Second thing is, most (if not all) readonly restrictions would probably concern either selected editable / root or whole editor. Latter one is easily accessed by `editor.isReadOnly`, but for the former one to have same accessibility, I've introduced another binding is `model.document.selection` component, that is `selectedRootIsReadOnly` observable (I am also thinking about adding a facade in editor to have same ease of retrieving information `editor.selectedRootIsReadOnly`, but at this moment I don't know if it's gonna be helpful.

Please note that at this moment, the contract is exactly the same (no breaking changes) as both `selectedRootIsReadOnly` and `document / rootElement.isReadOnly` depends solely on `editor.isReadOnly`.

2. Have some setup of roots we want to disable in `multiRootEditor` (currently, since I don't know how we want to pass that information, I haven't made any changes to the structure, but probably I'd recommend to have `multiRootEditor.sourceElements` to store an object with both `HTMLElement` and `isReadOnly` fields).  

Now, having some information which root we want to disable, we should rebind existing bindings from point one, that is rebind `rootElement.isReadOnly` and `editableElement.isReadOnly` to also include our additional information into bindings. That way, the readOnly restriction depends both of `editor.isReadOnly` and our new restriction in "mutliRootEditor". What's more important, it's completely invisible for other editors & plugins.

Next step that is not entirely covered in this PoC is changing behaviour of each plugin to either depend on `readOnly` on whole editor (eg. in case of internet shutdown) or `readOnly` on particular root (eg. in case of drag&drop widget in given root). Whatever is the case, having both `editor.isReadOnly` and either `editor.selectedRootIsReadOnly` or `model.document.selection.selectedRootIsReadOnly` make the change really easy.

File: packages/ckeditor5-engine/src/model/rootelement.ts [hunk 1]
```diff
@@ -10,11 +10,12 @@
 import Element from './element';
 
 import type Document from './document';
+import { EmitterMixin, ObservableMixin } from '@ckeditor/ckeditor5-utils';
 
 /**
  * Type of {@link module:engine/model/element~Element} that is a root of a model tree.
  */
-export default class RootElement extends Element {
+export default class RootElement extends ObservableMixin( EmitterMixin( Element ) ) {
```
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-engine/src/model/rootelement.ts:18]: This question will become obsolete if we move towards `MultiRootEditor`. See later comments.

File: packages/ckeditor5-engine/src/model/rootelement.ts [hunk 2]
```diff
@@ -25,6 +26,13 @@ export default class RootElement extends Element {
 	 */
 	private readonly _document: Document;
 
+	/**
+	 * Whether the element is in read-write or read-only mode.
+	 *
+	 * @observable
+	 */
+	declare public isReadOnly: boolean;
```
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-engine/src/model/rootelement.ts:34]: This question will become obsolete if we move towards `MultiRootEditor`. See later comments.

**Review by niegowski — CHANGES_REQUESTED** [review]
What we have today on master:

* Visual representation of the read-only mode
  * `view.Document#isReadOnly` is bound to `Editor#isReadOnly` (in `Editor`)
  * `view.EditableElement#isReadOnly` is bound to `view.Document#isReadOnly` (in `view.EditableElement`)
  * `contenteditable` attribute is bound to `view.EditableElement#isReadOnly` (in `view.View`)
* Behavior of the read-only mode - modifying the content (user actions in the editor)
  * Commands are disabled when `editor.isReadOnly` is set (in `Command`)
  * Some features check `Editor#isReadOnly` to block early before user action could trigger any command.

In this PR this looks like mixed with some additional flags that do not match the above flow. The assumption that some command should be blocked if the `DocumentSelection` is inside some blocked root seems invalid because many commands implement the possibility to specify target selection/range/position to perform some modifications (it does not always have to be a `DocumentSelection`).

**Review by scofalik — CHANGES_REQUESTED** [review]
TL;DR available below. For explanation, read full post. @niegowski @kbinieda please read full.

---

I think that the solution went in a good direction but we will need a bit more changes in regards to commands to have a nice API and commands behavior scheme.

AFAICS, we have several types of commands.

1. Commands that base only on document selection (e.g. bold).
1. Commands that change the model but do not base on the selection (e.g. insert paragraph).
1. Commands that sometimes base on selection and sometimes do not (e.g. table resize).
1. Commands that do not change the model (`affectsData = false` e.g. export to PDF).
1. Commands with other custom logic when they are enabled or not.

Currently, everything revolves around `Command#isEnabled`. But things will be more complicated now, so we need to decouple some behaviors.

First, `Command#isEnabled` should represent whether the command can be executed with default params (including default selection, that is document selection).

"Default params" is pretty vague, and it may mean different things, and sometimes there might not be "default params". But for now, let's think about simple commands like bold (depends just on the selection) or insert image (there's no default param for an image, but it doesn't matter what image you want to insert, so the command depends just on the selection).

Of course, these commands depend on editor readonly state too. Additionally, they should depend on whether the root is readonly or not (and here we have "depends on the selection"). Finally, they can also be disabled by some custom outside conditions, e.g. related to permissions.

I believe that all of this nicely applies to `isEnabled` and it still makes sense to say, that it represents whether the command can be executed with default params and "at the document selection".

I believe that currently there are two main "consumers" for `Command#isEnabled`. First is UI. Second is the check that prevents executing command if it is disabled.

For UI, we are interested in the "default params" execution. Think about toolbar buttons. Now they are linked with `Command#isEnabled`. It represents the regular command execution. It means "can I make this action right now?". It includes the general things (is editor readonly? did some other code disabled the command e.g. due to permissions?). But since it represents "calling it right now", in this case, `isEnabled` should include check for the current document selection.

_Side note: I actually believe that `isEnabled` is a simplification and a bit misleading property, at least in some cases, like commands that can be called for current document selection or called on a different range. Or with accept suggestion, where depending on which suggestion you accept, the command could be enabled or disabled. Many times, binding with `isEnabled` makes sense but not always._

For the second consumer -- the execute check -- it should check the general things but it should also check the actual call parameters. In this case, it should not include the check for the current document selection.

And we have a problem, because sometimes `isEnabled` should include the selection and sometimes not.

How to solve this problem?

First, again, let's summarize what interests us in the space of "command enablement":

1. Editor readonly state.
2. Are there other things that want to disable the command? (`Command#forceDisabled`).
3. Selection position.
    1. Check schema.
    2. Is current place (root) readonly?  
4. But our users can also have callbacks set directly on `change:isEnabled` event.
    1. These can be general callbacks.
    2. But, what's worst, there's nothing that stops the user from checking current selection in such callback. However, if they do it, then we should assume that their command always work on the default selection. 

In general, there's no way to solve this without at least some minor breaking changes.

The ideal solution would be to have:

1. `Command#isEnabled` that only has general states (editor `isReadonly`, command `forceDisabled`).
2. `Command#canBeExecuted( params )` where `params` is empty or is the same params that can be passed to `execute()`. The method should return `true` or `false` based on these params. In the most basic scenario, the method would not use any parameters and only make schema selection check + root readonly check. Or just return `true` if the command does not affect data. For most commands, `params` will not change whether a command can be executed or not, as it will depend only on the current selection.
3. "`canExecute`" that would combine `isEnabled` and `canBeExecuted()` without any params (so it will just check the current selection if needed). This is something that UI should be bound with.
4. On execute, we should check `isEnabled` + `canBeExecuted( params )` where params are params passed to `execute()` call.

However, it would be great to make as few changes as possible. This means that UI should be left as is. And `isEnabled` should include `canBeExecuted()`.

Then, on execute, we can check: `Editor#isReadOnly` + `Command#_disableStack.size` + canBeExecuted( params )`.

The only danger is that some integrators have some custom callbacks for `change:isEnabled`. It would be best to forbade that. Integrators should use `forceDisabled()` instead. The question is - can it be always used - but I believe so. It will just require to re-write some code on integrator's side, but it should be relatively easy. I believe that many times it looks like this:

```js
command.on( 'change:isEnabled', () => {
    if ( someProp ) {
        evt.returnValue = false;
        evt.stop();
    }
} );

// later on:
someProp = true;
command.refresh();
```

So, instead, you will have it even simpler:

```js
someProp = true;
command.forceDisabled();
```

Of course, `someProp` could be bind or something and `command.refresh()` could be in a callback but it doesn't change anything.

---

TL;DR

In summary. My proposal is as follows:

1. We introduce `Command#canBeExecuted( params? ): boolean` that returns whether the command can be executed with the default params (or with given `execute()` call params if they are set).
2. By default `Command#canBeExecuted()` will simply return `model.isSelectableEditable( this.editor.model.document.selection );`. This will be fine for 90% of commands.
3. `Command#isEnabled` = `Editor#isReadOnly` + `Command#_disableStack.size` + `Command#canBeExecuted()`. It should be refreshed on model `change` event (and of course when `Editor#isReadOnly` changes and `_disableStack.size` changes).
4. The "on execute" callback that currently checks `Command#isEnabled` should check: `Editor#isReadOnly` + `Command#_disableStack.size` + `Command#canBeExecuted( params )`.
5. We add documentation to `Command#isEnabled` that you cannot add callbacks to it and you should use `Command#forceDisabled()` instead. We write that in changelog and mark it as a breaking change (major or minor? I think minor).
6. Special commands (like the ones modified in this PR) will need to overwrite `canBeExecuted()` to include `params` check (if they are passed). Alternatively, they can make `canBeExecuted()` to always return `true` and then make extra checks directly inside `execute()`.

File: packages/ckeditor5-mention/src/mentioncommand.ts [hunk 3]
```diff
@@ -51,6 +51,13 @@ import type { MentionAttribute } from './mention';
  *	```
  */
 export default class MentionCommand extends Command {
+	public constructor( editor: Editor ) {
+		super( editor );
+
+		// Since this command may pass range in execution parameters, it should be checked directly in execute block.
+		this.stopListening( this.editor.model.document.selection, 'change' );
```
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-mention/src/mentioncommand.ts:58]: But it will not matter if we introduce changes as I proposed in the summary.

File: packages/ckeditor5-table/src/tablecolumnresize/tablecolumnresizeediting.ts [hunk 4]
```diff
@@ -471,6 +471,11 @@ export default class TableColumnResizeEditing extends Plugin {
 		const editor = this.editor;
 		const modelTable = editor.editing.mapper.toModelElement( target.findAncestor( 'figure' )! )!;
 
+		// Do not resize if table model is in non-editable place.
+		if ( !editor.model.isSelectableEditable( modelTable ) ) {
+			return;
+		}
+
```
Comment (scofalik) [inline packages/ckeditor5-table/src/tablecolumnresize/tablecolumnresizeediting.ts:478]: How about image resize? It doesn't need any changes?

File: packages/ckeditor5-paragraph/src/paragraphcommand.ts [hunk 5]
```diff
@@ -52,7 +59,14 @@ export default class ParagraphCommand extends Command {
 		const document = model.document;
 
 		model.change( writer => {
-			const blocks = ( options.selection || document.selection ).getSelectedBlocks();
+			const selection = options.selection || document.selection;
+
+			// Don't execute command if selection is in non-editable place.
+			if ( !model.isSelectableEditable( selection ) ) {
+				return;
+			}
```
Comment (scofalik) [inline packages/ckeditor5-paragraph/src/paragraphcommand.ts:67]: These lines (62-67) should be outside of `model.change()` so that it is not called if not needed.

**Comment by scofalik** [comment]
There's one more thing. I don't believe that commands now listen to roots changing their readonly status. That is, if I have a selection in root, I don't click anything and I don't make any action and then the root changes to read-only, then the command will not refresh now. Am I correct or I missed something? Since we moved the code to the multi-root editor, I am afraid that the multi-root editor will need to refresh all commands when the root starts/stops being readonly.

**Comment by scofalik** [comment]
When it comes to undo, AFAICS, each operation is reversed and transformed one after another. After the operation is transformed we could check if some of the resulting operations try to do changes in the readonly root and change them to `NoOperation`s.

Each operation will have to be treated separately as they have different parameters (sometimes position, sometimes two positions, sometimes a range). We need to check if all parameters operate in editable space (we can check it using `model.isSelectableEditable` as in other places).

We could try something like this but it will require complex and thorough testing. Thankfully, that is an edge case, as the root would have to become non-editable as the editor is initialized, which shouldn't be a common thing.

```javascript
for ( const operation of reversedOperations ) {
	if ( !isInsideEditablePlace( operation ) ) {
		operation = new NoOperation( operation.baseVersion );
	}
	
	// Before applying, add the operation to the `undoingBatch`.
	undoingBatch.addOperation( operation );
	model.applyOperation( operation );

	document.history.setOperationAsUndone( operationToUndo, operation );
}
```

File: packages/ckeditor5-clipboard/src/dragdrop.ts [hunk 6]
```diff
@@ -244,6 +244,13 @@ export default class DragDrop extends Plugin {
 		this.listenTo<ViewDocumentDragStartEvent>( viewDocument, 'dragstart', ( evt, data ) => {
 			const selection = modelDocument.selection;
 
+			// Don't drag from non-editable place.
+			if ( !model.isSelectableEditable( selection ) ) {
+				data.preventDefault();
+
+				return;
+			}
```
Comment (scofalik) [inline packages/ckeditor5-clipboard/src/dragdrop.ts:252]: Is there any specific reason to disable dragging when the root is readonly? AFAIR, it is not disabled currently.

File: packages/ckeditor5-engine/src/controller/editingcontroller.ts [hunk 7]
```diff
@@ -111,7 +111,11 @@ export default class EditingController extends ObservableMixin() {
 		// Whenever model document is changed, convert those changes to the view (using model.Document#differ).
 		// Do it on 'low' priority, so changes are converted after other listeners did their job.
 		// Also convert model selection.
-		this.listenTo<DocumentChangeEvent>( doc, 'change', () => {
+		this.listenTo<DocumentChangeEvent>( doc, 'change', evt => {
+			if ( evt.name === 'change:isReadOnly' ) {
+				return;
+			}
```
Comment (scofalik) [inline packages/ckeditor5-engine/src/controller/editingcontroller.ts:117]: It's a major problem to have a `change` event fired by a class "naturally", i.e. the class is not an observable. It's painful and we will lose the symmetry with the view but `change` event is too widely spread to have it this way. So, we cannot make `model.Document` an observable. Please propose a different solution.

File: packages/ckeditor5-undo/src/basecommand.ts [hunk 8]
```diff
@@ -203,6 +206,13 @@ export default abstract class BaseCommand extends Command {
 
 			// After reversed operation has been transformed by all history operations, apply it.
 			for ( const operation of reversedOperations ) {
+				console.log( 'ELO ', operation.type );
+				// Do not apply any operation on non-editable space.
+				if ( !this.editor.model.isSelectableEditable( operation.target() ) ) {
+					console.log( 'NOPE ', operation.type );
+					return;
+				}
+
```
Comment (scofalik) [inline packages/ckeditor5-undo/src/basecommand.ts:215]: Missed console logs.

File: packages/ckeditor5-undo/src/basecommand.ts [hunk 9]
```diff
@@ -203,6 +206,13 @@ export default abstract class BaseCommand extends Command {
 
 			// After reversed operation has been transformed by all history operations, apply it.
 			for ( const operation of reversedOperations ) {
+				console.log( 'ELO ', operation.type );
+				// Do not apply any operation on non-editable space.
+				if ( !this.editor.model.isSelectableEditable( operation.target() ) ) {
+					console.log( 'NOPE ', operation.type );
+					return;
```
Comment (scofalik) [inline packages/ckeditor5-undo/src/basecommand.ts:213]: Why `return;` instead of returning `NoOperation` as I suggested in my previous review? Were there any problems with my approach?

**Review by scofalik — CHANGES_REQUESTED** [review]

**Comment by scofalik** [comment]
Plus, I am not sure whether it reacts properly to roots changing their readonly state, please confirm that it's okay or add this.

File: packages/ckeditor5-editor-multi-root/src/multirooteditor.ts [hunk 10]
```diff
@@ -185,6 +211,11 @@ export default class MultiRootEditor extends DataApiMixin( Editor ) {
 				}
 			}
 		} );
+
+		this.listenTo<ModelIsSelectableEditableEvent>( this.model, 'isSelectableEditable', ( evt, [ selectable ] ) => {
+			evt.return = this.isSelectableEditable( selectable );
+			evt.stop();
```
[the remark this answers is not in the document]
Comment (kbinieda) [reply packages/ckeditor5-editor-multi-root/src/multirooteditor.ts:228]: Not really, that is, what I want to achieve is to not call default method (which is binded to document.isReadOnly).

File: packages/ckeditor5-table/src/tablecolumnresize/tablecolumnresizeediting.ts (hunk 4 above)
Comment (kbinieda) [reply packages/ckeditor5-table/src/tablecolumnresize/tablecolumnresizeediting.ts:478]: It seems not needed.

File: packages/ckeditor5-editor-multi-root/src/multirooteditor.ts [hunk 11]
```diff
@@ -481,6 +526,138 @@ export default class MultiRootEditor extends DataApiMixin( Editor ) {
...
+			rootName = selectable.root.rootName;
+		} else if ( selectable instanceof Node ) {
+			rootName = selectable.root.rootName;
+		} else if ( isIterable( selectable ) ) {
+			for ( const range of selectable as Iterable<Range> ) {
+				rootName = range.start.root.rootName;
+
+				break; // Since all ranges should have same root we can only check the first element.
+			}
+		}
```
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-editor-multi-root/src/multirooteditor.ts:652]: Maybe we should modify the decorated model method to normalize selectable same as `insertContent` does.

File: packages/ckeditor5-editor-multi-root/src/multirooteditor.ts [hunk 12]
```diff
@@ -122,6 +136,28 @@ export default class MultiRootEditor extends DataApiMixin( Editor ) {
...
+				const viewRange = this.editing.view.createRangeIn( viewRoot );
+
+				for ( const viewItem of viewRange.getItems() ) {
+					if ( viewItem.is( 'editableElement' ) ) {
+						viewItem.unbind( 'isReadOnly' );
+						viewItem.isReadOnly = value;
+					}
+				}
+			} );
+		} );
```
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-editor-multi-root/src/multirooteditor.ts:159]: View document roots collection is bound to model document roots, and model document roots are never removed.

File: packages/ckeditor5-core/src/command.ts [hunk 13]
```diff
@@ -89,29 +101,39 @@ export default class Command extends ObservableMixin() {
...
 
 		this.decorate( 'execute' );
 
-		// By default every command is refreshed when changes are applied to the model.
+		// By default, every command is refreshed when changes are applied to the model.
 		this.listenTo( this.editor.model.document, 'change', () => {
 			this.refresh();
 		} );
 
+		this.listenTo<ObservableChangeEvent<boolean>>( editor, 'change:isReadOnly', () => {
```
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-core/src/command.ts:114]: :+1:.

File: packages/ckeditor5-core/src/editor/editor.ts [hunk 14]
```diff
@@ -304,6 +304,10 @@ export default abstract class Editor extends ObservableMixin() {
 
 		this.model = new Model();
 
+		this.on( 'change:isReadOnly', () => {
+			this.model.document.isReadOnly = this.isReadOnly;
```
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-core/src/editor/editor.ts:308]: :+1:.

File: packages/ckeditor5-clipboard/src/dragdrop.ts [hunk 15]
```diff
@@ -480,11 +489,15 @@ export default class DragDrop extends Plugin {
 			// In Firefox this is not needed. In Safari it makes the whole editable draggable (not just textual content).
 			// Disabled in read-only mode because draggable="true" + contenteditable="false" results
 			// in not firing selectionchange event ever, which makes the selection stuck in read-only mode.
-			if ( env.isBlink && !editor.isReadOnly && !draggableElement && !viewDocument.selection.isCollapsed ) {
+			if ( env.isBlink && !draggableElement && !viewDocument.selection.isCollapsed ) {
 				const selectedElement = viewDocument.selection.getSelectedElement();
 
 				if ( !selectedElement || !isWidget( selectedElement ) ) {
-					draggableElement = viewDocument.selection.editableElement;
+					const editableElement = viewDocument.selection.editableElement;
```
[the remark this answers is not in the document]
Comment (scofalik) [reply packages/ckeditor5-clipboard/src/dragdrop.ts:496]: I believe that current implementation is proper and more precise. We want to avoid setting draggable="true and contenteditable="false" on the same element. It checks exactly that.

File: packages/ckeditor5-editor-multi-root/src/multirooteditor.ts [hunk 16]
```diff
@@ -231,11 +229,21 @@ export default class MultiRootEditor extends DataApiMixin( Editor ) {
...
-			evt.return = isEditable;
+		// Overwrite `Model#canEditAt()` decorated method.
+		// Check if the provided selection is inside a read-only root. If so, return `false`.
+		this.listenTo<ModelCanEditAtEvent>( this.model, 'canEditAt', ( evt, [ selection ] ) => {
+			// Skip empty selections.
+			if ( !selection.getFirstPosition() ) {
+				return;
+			}
+
+			const rootName = selection.getFirstPosition()!.root.rootName!;
```
Comment (niegowski) [inline packages/ckeditor5-editor-multi-root/src/multirooteditor.ts:240]: What if there are multiple roots coming from move to graveyard? Should we check all roots?

File: packages/ckeditor5-engine/src/model/model.ts [hunk 17]
```diff
@@ -1128,7 +1146,15 @@ function normalizeSelectable(
 	}
 
 	if ( selectable instanceof Node ) {
-		return new ModelSelection( selectable, placeOrOffset! );
+		if ( placeOrOffset ) {
```
Comment (niegowski) [inline packages/ckeditor5-engine/src/model/model.ts:1149]: what about `=== 0`? Maybe this should check if its is `!== null`?

**Review by niegowski — APPROVED** [review]
